### ISSUE TITLE
Intel graphics support

### ISSUE THREAD

**Issue by claell** [virtual-programming/dirtshowdown-linux#10]
Hello,
first of all thank you for porting the game. After activating it and installing I got an error message at start. Later I noticed the warning that there is currently no support for Intel graphics.

Will the support be added later and is there an ETA?

**Comment by vpltd-peterm** [comment]
Support for this is entirely down to development of the Intel driver. We do not specifically stop the game running on any particular driver, but it must supply the features we require.

**Comment by claell** [comment]
So the driver needs to support OpenGL 4.1 first?

**Comment by vpltd-peterm** [comment]
Correct

**Comment by claell** [comment]
Alright, thanks.

**Comment by claell** [comment]
I don't know much what Mesa is, but if I had Mesa installed that implements OpenGL 4.1 (like someone here: http://steamcommunity.com/app/201700/discussions/0/520518053446469553/) why does the game still not start?

**Comment by vpltd-peterm** [comment]
Mesa is an open source graphics driver, and this is what the Intel graphics driver is based on. GL 4.1 support in Mesa is quite a recent development, and it is not fully there yet.

**Comment by claell** [comment]
Mesa claims that it is and there seem also to be versions which support OpenGL 4.1. So does the game not recognize Mesa and only looks for the normal Intel driver?

**Comment by vpltd-kgalaj** [comment]
We generally aim to support Mesa, if we don't have to go as far as to write missing OpenGL futures for the driver in order to do so. So, even if now our games don't generally work on Mesa drivers, we hope they will in the future, as new driver versions add support for OpenGL features the games need to run.

Please advise if you have the latest update to the game, and also what happens, exactly, when you launch the game when using Mesa driver. The contents of ~/.local/share/vpltd/dirt/eon.txt might help a lot with determining why it didn't work on your machine - if it's something we can fix on our side, we'll do so.

**Comment by claell** [comment]
I just installed MESA 11.1 which claims to support OpenGL 4.1 (http://mesamatrix.net/), at least this is how it looks to me. Although glxinfo | grep "OpenGL version" shows:

OpenGL version string: 3.0 Mesa 11.1.0-devel (git-f8a9388 2015-09-01 vivid-oibaf-ppa)

So maybe my system is not using the Mesa driver?

eon.txt:

```
[thread ffffffff][I][0]: Log system initialised
[thread ffffffff][I][5]: Default allowed number of simultaneously open files in the system is 2448.
[thread ffffffff][I][5]: Current memlock limit is 65536.
[thread ffffffff][I][5]: setrlimit(RLIMIT_MEMLOCK) failed, (1) Operation not permitted
[thread ffffffff][I][5]: setup_teb_register() - using LDT entry 13
[thread 00000001][I][20]: Forcing OpenGL Core context usage
[thread 00000001][I][21]: Creating OpenGL Core context failed
[thread 00000001][E][3140]: PrepareOpenGLDisplayInformation() failed!!
```

**Comment by vpltd-kgalaj** [comment]
Well, between "Forcing OpenGL Core context usage" and "Creating OpenGL Core context failed", we're basically setting up typical SDL OpenGL 4.1 context attributes:

SDL_GL_DOUBLEBUFFER 1
SDL_GL_ACCELERATED_VISUAL 1
SDL_GL_RED_SIZE, SDL_GL_GREEN_SIZE and SDL_GL_BLUE_SIZE 8 for 24-bit display modes and 10 for 30-bit display modes,
SDL_GL_CONTEXT_PROFILE_MASK SDL_GL_CONTEXT_PROFILE_CORE
SDL_GL_CONTEXT_MAJOR_VERSION 4
SDL_GL_CONTEXT_MINOR_VERSION 1
SDL_GL_MULTISAMPLEBUFFERS 0
SDL_GL_MULTISAMPLESAMPLES 0

and calling SDL_GL_CreateContext(), letting well-known and supported SDL library create OpenGL context. So, if that fails, it's somewhere below us, and likely either between driver and SDL, or it may be failing because the display you're trying to set it up on has only some very strange display modes supported. You could try launching it on the most common 1280x1024 or 1024x768 ARGB8 display mode to see if that might be it.

**Comment by claell** [comment]
Thanks for the answer. I am not really sure, whether my driver even supports OpenGL 4.1 since glxinfo | grep "OpenGL" delivers:

```
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Haswell Desktop 
OpenGL core profile version string: 3.3 (Core Profile) Mesa 11.1.0-devel (git-f8a9388 2015-09-01 vivid-oibaf-ppa)
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 11.1.0-devel (git-f8a9388 2015-09-01 vivid-oibaf-ppa)
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 11.1.0-devel (git-f8a9388 2015-09-01 vivid-oibaf-ppa)
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:
```
Or is everything running through Mesa now and the OpenGL 3.3 version supported by my driver does not matter?

 How do I launch the game with the proposed modes?

**Comment by vpltd-kgalaj** [comment]
It looks like it says it supports OpenGL 3.3 (core profile version string, not 4.1). Perhaps it supports almost all OpenGL 4.1 features through OpenGL extensions, but we really need practically all OpenGL 4.1 can offer (and a good deal beyond), not only subsets of functionality through extensions.

**Comment by claell** [comment]
Yes I understand that. What I do not understand is what the difference between Mesa and the Intel driver is. Because it seems that Mesa already fully supports OpenGL 4.1 and only the Intel driver does not. So I thought what if I use Mesa instead. But maybe I did not understand, what Mesa is.

**Comment by vpltd-peterm** [comment]
Put simply, Mesa is a driver framework, which Intel use to implement their driver. There are drivers using Mesa that support GL 4.1 now, but the Intel driver is not one of them.

**Comment by claell** [comment]
Alright. Thanks for the explanation.

**Comment by vpltd-peterm** [comment]
Closing as not a bug in the game itself

**Comment by cplt** [comment]
On Intel Iris Pro 6200 (and probably any other somewhat modern Intel GPU) with Mesa 11.0.5 and using MESA GL and GLSL version overrides (e.g. MESA_GL_VERSION_OVERRIDE=4.2 MESA_GLSL_VERSION_OVERRIDE=420) to pretend the driver is fully OpenGL 4.1 or 4.2 compliant, the game actually starts and the graphics look fine, except that half of the screen is obscured by a black triangle (with the triangle being in the upper right corner). Other than that, it is playable and the frame rate is okay too. This issue does not look like it is caused by the missing OpenGL extensions to me.

**Comment by vpltd-peterm** [comment]
This indicates the final render to screen operation is not drawing the complete quad it is supposed to.  This is a driver bug.

**Comment by TruboXL** [comment]
I hate to comment this but this ugly hack can force the game to run the currently "OpenGL 3.3 capable" Intel GPU **and also fix the invisible tires issue #4** !
Credits: https://www.reddit.com/r/linux/comments/41s1ym/why_is_intel_taking_so_much_time_to_finish_the/cz5w273

In the launch option for Dirt Showdown, add this line:
>MESA_GL_VERSION_OVERRIDE=4.1 MESA_GLSL_VERSION_OVERRIDE=410 %command%
>I don't include GL_ARB_compute_shader is OpenGL 4.3 feature and will crash the game if set to Medium settings and up

That said, Intel GPU is capable running OpenGL 4+ and drivers are still being worked on...
But devs please take a look why the hack fix #4 ~
